### PR TITLE
Alerting: Create folder for alerting when start from the scratch

### LINKED ISSUE #45303: Alerting: improve onboarding by having a default folder for alerts
New Grafana instances do not have any actual folders where alerts can be stored; perhaps a migration in the backend to provision an alerts folder if there isn't one – this would help HG users and Grafana OSS users to create a new alert without having to think about dashboard folders.

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
When a brand new Grafana starts, there are no folders and therefore, in order to create the first alert, users will have to create a new folder. This can be confusing for first-time users that are not very familiar with the conception of folders in Grafana.
This PR adds a migration check whether an organization has any folders and if it does not have any folder, creates a new one with the name "General Alerting". The folder is created only for organizations for which the alerting is enabled.

**Which issue(s) this PR fixes**:
Fixes  https://github.com/grafana/grafana/issues/45303

### CODE REVIEW

File: pkg/services/sqlstore/migrations/ualert/permissions.go [hunk 1]
```diff
@@ -40,9 +43,14 @@ type dashboardAcl struct {
 	Updated time.Time
 }
 
+type folderHelper struct {
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/permissions.go:46]: Do we need an extra level of abstraction just for this?

File: pkg/services/sqlstore/migrations/ualert/permissions.go [hunk 2]
```diff
@@ -279,3 +287,18 @@ func (m *migration) getACL(orgID, dashboardID int64) ([]*dashboardAcl, error) {
 	err = m.sess.SQL(rawSQL, orgID, dashboardID).Find(&result)
 	return result, err
 }
+
+// getOrgsThatHaveFolders returns a list of orgID that have at least one folder
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/permissions.go:291]: This method returns a map (or "set"), not a list.

File: pkg/services/sqlstore/migrations/ualert/permissions.go [hunk 3]
```diff
@@ -279,3 +287,18 @@ func (m *migration) getACL(orgID, dashboardID int64) ([]*dashboardAcl, error) {
...
+func (m *folderHelper) getOrgsThatHaveFolders() (map[int64]struct{}, error) {
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/permissions.go:292]:
```suggestion
func (m *folderHelper) getOrgIDsThatHaveFolders() (map[int64]struct{}, error) {
```

File: pkg/services/sqlstore/migrations/ualert/ualert.go [hunk 4]
```diff
@@ -276,6 +276,11 @@ func (m *migration) Exec(sess *xorm.Session, mg *migrator.Migrator) error {
 			}
 		}
 
+		folderHelper := folderHelper{
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/ualert.go:279]: Again, not sure if this extra struct adds much value

File: pkg/services/sqlstore/migrations/ualert/ualert.go [hunk 5]
```diff
@@ -753,3 +758,58 @@ func getAlertFolderNameFromDashboard(dash *dashboard) string {
...
+	var rows []struct {
+		Id   int64
+		Name string
+	}
+
+	if err := sess.Table("org").Cols("id", "name").Find(&rows); err != nil {
+		return fmt.Errorf("failed to read the list of organizations: %w", err)
+	}
+
+	numberFoldersPerOrg, err := helper.getOrgsThatHaveFolders()
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/ualert.go:789]: I think the actual purpose of the method is not captured by the variable name and that can lead to confusion. This method does not return the number of folders per organization, but a set of IDs belonging to organizations that have at least one folder.

```suggestion
	orgsWithFolders, err := helper.getOrgIDsThatHaveFolders()
```

File: pkg/services/sqlstore/migrations/ualert/ualert.go [hunk 6]
```diff
@@ -753,3 +758,58 @@ func getAlertFolderNameFromDashboard(dash *dashboard) string {
...
+	if err != nil {
+		return fmt.Errorf("failed to count folders per organization: %w", err)
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/ualert.go:791]: Same than above, AFAIU the method doesn't count the number of folders per org, but the orgs with folders. Maybe that was the original purpose of the function but then it you changed it to better fit our needs?

File: pkg/services/sqlstore/migrations/ualert/ualert.go [hunk 7]
```diff
@@ -753,3 +758,58 @@ func getAlertFolderNameFromDashboard(dash *dashboard) string {
...
+	}
+
+	for _, row := range rows {
+		// if there is at least one folder in the organization. Skip adding the default folder
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/ualert.go:795]:
```suggestion
		// if there's at least one folder in the org or if alerting is disabled for that org, skip adding the default folder
```

File: pkg/services/sqlstore/migrations/ualert/ualert.go [hunk 8]
```diff
@@ -753,3 +758,58 @@ func getAlertFolderNameFromDashboard(dash *dashboard) string {
...
+		if _, ok := numberFoldersPerOrg[row.Id]; ok {
+			migrator.Logger.Debug("Skip adding default alerting folder because organization already has at least one folder", "org_id", row.Id)
+			continue
+		}
+		if _, ok := migrator.Cfg.UnifiedAlerting.DisabledOrgs[row.Id]; ok {
+			migrator.Logger.Debug("Skip adding default alerting folder because alerting is disabled for the organization ", "org_id", row.Id)
+			continue
+		}
+		folder, err := helper.getOrCreateGeneralFolder(row.Id)
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/ualert.go:804]: We're sure that no folder exists for this org. We can skip the query part and just creating the general folder using a separate method (e.g., `helper.createGeneralFolder(row.Id)`)

File: pkg/services/sqlstore/migrations/ualert/ualert.go [hunk 9]
```diff
@@ -753,3 +758,58 @@ func getAlertFolderNameFromDashboard(dash *dashboard) string {
...
+		if err != nil {
+			return fmt.Errorf("failed to create the default alerting folder for organization %s (ID: %d): %w", row.Name, row.Id, err)
+		}
+		migrator.Logger.Info("created the default folder for alerting", "org_id", row.Id, "folder_name", folder.Title, "folder_uid", folder.Uid)
```
Comment (santihernandezc) [inline pkg/services/sqlstore/migrations/ualert/ualert.go:808]: All titles would be the same, we are using the `GENERAL_FOLDER` constant for this.

```suggestion
		migrator.Logger.Info("created the default folder for alerting", "org_id", row.Id, "folder_uid", folder.Uid)
```

**Review by santihernandezc — CHANGES_REQUESTED** [review]

File: pkg/services/sqlstore/migrations/ualert/ualert.go (hunk 9 above)
Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/ualert.go:808]: right, but those who may look into logs do not have to know about that constant.

File: pkg/services/sqlstore/migrations/ualert/ualert.go (hunk 6 above)
Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/ualert.go:791]: It used to be count but then I simplified it even more

File: pkg/services/sqlstore/migrations/ualert/permissions.go (hunk 1 above)
Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/permissions.go:46]: I did that to share the code for creating a folder between two migrations.

File: pkg/services/sqlstore/migrations/ualert/permissions.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/permissions.go:291]: a list, as well as a collection, is a more general description of a map or set.

File: pkg/services/sqlstore/migrations/ualert/ualert.go (hunk 4 above)
Comment (yuri-tceretian) [reply pkg/services/sqlstore/migrations/ualert/ualert.go:279]: if you have better idea, I will be glad to consider it.

File: pkg/services/sqlstore/migrations/ualert/permissions.go (hunk 1 above)
Comment (santihernandezc) [reply pkg/services/sqlstore/migrations/ualert/permissions.go:46]: I can think of two simple options:
1. Just use `*migration`.
2. Use `folderHelper` but change the methods' signature to make it clear that we're referring to this new struct.
```go
func (m *folderHelper) getOrCreateGeneralFolder(orgID int64) (*dashboard, error)
vs
func (h *folderHelper) getOrCreateGeneralFolder(orgID int64) (*dashboard, error) {
```

**Review by santihernandezc — APPROVED** [review]
LGTM, all important changes have been addressed, I left a comment but it's a readability thing. Great job!

**Pull request merged**